### PR TITLE
Make profile dates match style of website

### LINKED ISSUE #2590: Unify "Joined.." text format + tooltips on hover with the web
The website has text like `Joined December 2014`, while Lazer has `Joined 3 years ago`. Also the tooltips in Lazer seem lacking compared to the web.
I'm looking into fixing this, ~but I don't really know how I should get `user.JoinDate.Month` into words like "December". Is there already something implemented to handle conversions like this?~ [Nvm, just did it with string.Format()](https://i.imgur.com/Jm8BHUX.png?1). Now I'll try to make tooltips the same, though it looks like I might have to modify some things to change the way date is shown in the tooltips, idk.

Website:

Lazer:

Also, as visible on the screenshot from the client, I have variable texts in my system's language, is there some way to go 100% English? And should I really change Lazer's way of displaying join date to be like web? Shouldn't it be the other way around?

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58578961-unify-joined-text-format-tooltips-on-hover-with-the-web?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by HoutarouOreki**
Made to resolve #2590

Got as close as I could, but tooltips can't be formatted like on the web, as far as I'm aware. Also I have no idea what I'm doing, so please revise these changes and tell me what I did wrong and what I can make better.

---

Now more localised and special than before.

### CODE REVIEW

File: osu.Game/Graphics/DrawableDate.cs [hunk 1]
```diff
@@ -13,13 +13,59 @@ namespace osu.Game.Graphics
...
+            tooltipFormat = "";
+        }
+
+        public DrawableDate(string dateFormat, DateTimeOffset date)
+        {
+            AutoSizeAxes = Axes.Both;
+            Font = "Exo2.0-RegularItalic";
+
+            this.date = date.ToLocalTime();
+            // set a date format for later from an argument
```
Comment (smoogipoo) [inline osu.Game/Graphics/DrawableDate.cs]: don't need this type of comment - it's self explanatory

File: osu.Game/Graphics/DrawableDate.cs [hunk 2]
```diff
@@ -13,13 +13,59 @@ namespace osu.Game.Graphics
...
             this.date = date.ToLocalTime();
+            // if date's format is not specified, set it to an empty string,
+            // so that later we will know to humanize it
+            dateFormat = "";
+            // if tooltip's format is not specified, set it to an empty string
+            // so later we will know to default to a default time format
+            tooltipFormat = "";
+        }
+
+        public DrawableDate(string dateFormat, DateTimeOffset date)
```
Comment (smoogipoo) [inline osu.Game/Graphics/DrawableDate.cs]: Instead of 3 new constructors, how about change the one that existed previously to `public DrawableDate(DateTimeOffset date, string dateFormat = null, string tooltipFormat = null)`

With comment on dateformat: "The string to format the date text with. May be null if the humanized format should be used."
And on tooltipFormat: "The string to format the tooltip text with. May be null if the default format should be used."

The comments in the ctor would be unnecessary then.

File: osu.Game/Graphics/DrawableDate.cs [hunk 3]
```diff
@@ -13,13 +13,59 @@ namespace osu.Game.Graphics
...
             AutoSizeAxes = Axes.Both;
             Font = "Exo2.0-RegularItalic";
 
             this.date = date.ToLocalTime();
+            // if date's format is not specified, set it to an empty string,
+            // so that later we will know to humanize it
+            dateFormat = "";
+            // if tooltip's format is not specified, set it to an empty string
+            // so later we will know to default to a default time format
+            tooltipFormat = "";
```
Comment (smoogipoo) [inline osu.Game/Graphics/DrawableDate.cs]: Setting to empty string is unnecessary.

`string` is null by default, so you should use `Text = string.IsNullOrEmpty(dateFormat) ? A : B`

File: osu.Game/Graphics/DrawableDate.cs [hunk 4]
```diff
@@ -13,13 +13,59 @@ namespace osu.Game.Graphics
     public class DrawableDate : OsuSpriteText, IHasTooltip
     {
         private readonly DateTimeOffset date;
+        private readonly String dateFormat;
+        private readonly String tooltipFormat;
```
Comment (smoogipoo) [inline osu.Game/Graphics/DrawableDate.cs]: `string` instead of `String`

File: osu.Game/Graphics/DrawableDate.cs [hunk 5]
```diff
@@ -58,8 +104,18 @@ private void updateTimeWithReschedule()
 
         public override bool HandleMouseInput => true;
 
-        private void updateTime() => Text = date.Humanize();
+        // if date's format is specified
+        private void updateTime() => Text = dateFormat != "" ?
```
Comment (smoogipoo) [inline osu.Game/Graphics/DrawableDate.cs]: `string.IsNullOrEmpty` here and below, also too many comments, pretty self explanatory what all this does.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Graphics/DrawableDate.cs [hunk 6]
```diff
@@ -13,13 +13,21 @@ namespace osu.Game.Graphics
...
 
-        public DrawableDate(DateTimeOffset date)
+        public DrawableDate(DateTimeOffset date, string dateFormat = null, string tooltipFormat = null)
         {
             AutoSizeAxes = Axes.Both;
             Font = "Exo2.0-RegularItalic";
 
             this.date = date.ToLocalTime();
+            // The string to format the date text with.
+            // May be null if the humanized format should be used.
```
Comment (Aergwyn) [inline osu.Game/Graphics/DrawableDate.cs]: This should be doc on the constructor itself.

**Comment by peppy** [comment]
To be honest, I'd rather the formattable portions be `virtual` and we just have a set number of classes for each display format. I guarantee there won't be more than 2 or 3 and I don't want to have to copy paste format strings each time.

**Comment by HoutarouOreki** [comment]
Sorry, can I get an ELI5? I tried to understand but I might be retarded.

Should there be predefined strings?

Or is it supposed to be something like
```c#
protected virtual DrawableDateSomeFormat(text) => new DrawableDate(text, predefinedFormats)
```

If formattable portions should be virtual, then I guess something else?.. I don't really know what it means.

**Comment by HoutarouOreki** [comment]
Ok, it works. Not sure I got the code *100%* as you want it though.

File: osu.Game/Graphics/DrawableDate.cs [hunk 7]
```diff
@@ -58,8 +58,10 @@ private void updateTimeWithReschedule()
 
         public override bool HandleMouseInput => true;
 
-        private void updateTime() => Text = date.Humanize();
+        protected virtual string Format() => Text = date.Humanize();
 
-        public string TooltipText => date.ToString();
+        private void updateTime() => Format();
+
+        public virtual string TooltipText => string.Format($"{date:d MMMM yyyy H:mm \"UTC\"z}");
```
Comment (DrabWeb) [inline osu.Game/Graphics/DrawableDate.cs]: Format should be `MMMM d, yyyy h:mm tt \"UTC\"z` to match web.

image

File: osu.Game/Graphics/DrawableJoinDate.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        private readonly DateTimeOffset date;
+
+        public DrawableJoinDate(DateTimeOffset date) : base(date)
+        {
+            this.date = date;
+        }
+
+        protected override string Format() => Text = string.Format($"{date:MMMM yyyy}");
+
+        public override string TooltipText => string.Format($"{date:d MMMM yyyy}");
```
Comment (DrabWeb) [inline osu.Game/Graphics/DrawableJoinDate.cs]: Format should be `MMMM d, yyyy` to match web.

image

File: osu.Game/Graphics/DrawableDate.cs [hunk 9]
```diff
@@ -58,8 +58,10 @@ private void updateTimeWithReschedule()
 
         public override bool HandleMouseInput => true;
 
-        private void updateTime() => Text = date.Humanize();
+        protected virtual string Format() => Text = date.Humanize();
```
Comment (DrabWeb) [inline osu.Game/Graphics/DrawableDate.cs]: Return the string here, then when called in `updateTime` set `Text` to the return value.

File: osu.Game/Graphics/DrawableJoinDate.cs (hunk 8 above)
Comment (DrabWeb) [reply osu.Game/Graphics/DrawableJoinDate.cs]: This class should probably also handle the "Here since the beginning" text. As it is when hovering over that text in the profile overlay, no date is shown.

Comment (HoutarouOreki) [reply osu.Game/Graphics/DrawableJoinDate.cs]: 1: I guess I will have to check if there's some culture-specific formatting, 'cause the text varies language-to-language.

Comment (DrabWeb) [reply osu.Game/Graphics/DrawableJoinDate.cs]: It does right now, I'm saying `DrawableJoinDate` should instead.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 10]
```diff
@@ -357,16 +357,13 @@ private void loadUser()
 
             infoTextLeft.NewParagraph();
 
-            if (user.JoinDate.ToUniversalTime().Year < 2008)
-            {
-                infoTextLeft.AddText("Here since the beginning", boldItalic);
-            }
-            else
+            if (user.JoinDate.ToUniversalTime().Year >= 2008)
```
Comment (Aergwyn) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: Would it be possible to move "Joined" into `DrawableJoinDate` as well so you don't need to have this here anymore?

Comment (HoutarouOreki) [reply osu.Game/Overlays/Profile/ProfileHeader.cs]: The thing is, I forgot that "Here since the beginning" should be formatted in _lightText_, so this IF will be here anyway. "Joined" part of the text also doesn't give a tooltip on the web, so that would also mean de-unification.

**Comment by HoutarouOreki** [comment]
There's a problem with dates' formats being different across the languages. C#'s _.format()_ doesn't really have an equivalent to the web's Moment's format in this place there, as far as I'm aware.

**Comment by Aergwyn** [comment]
You may want to look into `CultureInfo.InvariantCulture` as a parameter then.
Also ppy/osu-web#3357 may be of interest while it's still WIP.

**Comment by HoutarouOreki** [comment]
Using `CultureInfo.InvariantCulture` as a parameter does exactly the opposite of what needs to be done, if I'm using it correctly. This misunderstanding is my fault, I didn't explain the problem thoroughly. When passing `InvariantCulture` as a parameter, month names are in English on my Polish system.

What I want to do, is have it automatically format the date according to system's culture/language, as osu-web does. For example, English version of osu-web has `MMMM d, yyyy`, and Polish has `d MMMM yyyy` for join dates' tooltips. Various cultures have different formats. English date format looks _extremely_ unsettling in Polish.

The closest to achieving the goal is `date.ToString("D", DateTimeFormatInfo)`, except that it puts a weekday in front of the date, so it's `dddd, MMMM dd, yyyy` for EN and `dddd, d MMMM yyyy` for PL. osu-web has no weekday there.

I feel like I shouldn't use regex to remove the `dddd` weekday..? Though I don't see any other solution. So should I? 💢

**Review by peppy — APPROVED** [review]

**Review by peppy — APPROVED** [review]
